snet_cli/utils.py

```diff
@@ -83,3 +85,50 @@ def type_converter(t):
         return web3.Web3.toChecksumAddress
     else:
         return str
+
+
+def _validate_path(path, entry_path):
+    return entry_path in path.parents
+
+
+def _add_next_paths(path, parent, entry_path_parents, walked_paths, next_paths):
+    with open(path) as f:
+        for line in f:
+            if line.strip().startswith("import"):
+                import_statement = "".join(line.split('"')[1::2])
+                if not import_statement.startswith("google/protobuf"):
+                    import_statement_path = PurePath(Path(parent.joinpath(import_statement)).resolve())
+                    if _validate_path(import_statement_path, entry_path_parents):
+                        if import_statement not in walked_paths:
+                            next_paths.append(import_statement_path)
+                    else:
+                        raise ValueError("Path must not be a parent of entry path")
+
+
+def walk_imports(model_path):
+    if os.path.isdir(model_path):
+        entry_path = Path.cwd().joinpath(model_path).resolve()
+    elif os.path.isfile(model_path):
+        entry_path = Path.cwd().joinpath(model_path).parent.resolve()
+    else:
+        raise IOError("Model directory must be a valid directory: {}".format(model_path))
+    walked_paths = set()
+    next_paths = []
+    for file in os.listdir(entry_path):
+        if file.endswith(".proto"):
+            next_paths.append(entry_path.joinpath(file))
+    while next_paths:
+        path = next_paths.pop()
+        if os.path.isfile(path):
+            parent = path.parent
+            _add_next_paths(path, parent, entry_path, walked_paths, next_paths)
+            walked_paths.add(path)
```

Comment:
i'd add to `walked_paths` when you add to `next_paths` (maybe better naming like `seen_paths` and `paths_to_walk` or something) because consider

A imports B
A imports C
B imports C

```
next_paths = [A]
walked_paths = []
path = next_paths.pop() == A => next_paths == []
add_next_paths(A) => next_paths == [B,C]
walked_paths.add(A) => walked_paths == [A]
path = next_paths.pop() == B => next_paths == [C]
add_next_paths(B) => next_paths == [C,C] *********************
walked_paths.add(B) => walked_paths == [A,B]
...
```

Reply:
Fixed